torch/distributed/pipeline/sync/checkpoint.py

```diff
@@ -302,32 +312,49 @@ def forward(
         rng_states: Deque[RNGStates],
         function: Function,
         input_atomic: bool,
-        *input: Tensor,
+        *inputs,
     ) -> Tensor:
         ctx.recomputed = recomputed
         ctx.rng_states = rng_states
 
         ctx.function = function
         ctx.input_atomic = input_atomic
-        ctx.save_for_backward(*input)
+        ctx.inputs = inputs
+        if input_atomic:
+            tensors = [inputs[0]]
+        else:
+            tensors = []
+            for input in enumerate(inputs):
+                if torch.is_tensor(input):
+                    tensors.append(input)
+        ctx.save_for_backward(*tensors)
 
         return phony
 
     @staticmethod
     def backward(ctx: Context, *grad_output: Tensor) -> Tuple[None, ...]:  # pragma: no cover
-        input = ctx.saved_tensors
-        input_leaf = tuple(x.detach().requires_grad_(x.requires_grad) for x in input)
+        inputs = ctx.inputs
+        inputs_leaf = tuple(x.detach().requires_grad_(x.requires_grad) if torch.is_tensor(x) else x for x in inputs)
+
+        # Get the device for the inputs from a tensor
+        device = None
+        for input in inputs:
+            if torch.is_tensor(input):
+                device = input.device
```

Comment:
currently device will be the device of the last input tensor - can we break when we populate device the first time? Is there any issue if they're on different devices?